components/ai_chat/resources/page/stories/components_panel.tsx

```diff
@@ -213,13 +233,72 @@ export default {
       const apiHasError = currentError !== mojom.APIError.None
       const shouldDisableUserInput = apiHasError || isGenerating
       const currentModel = MODELS.find(m => m.name === options.args.model)
-      const [inputText, setInputText] = useState('')
+      const [inputText, setInputText_] = useState('')
+      const [selectedActionType, setSelectedActionType] = useState<mojom.ActionType | undefined>(undefined)
+      const [isToolsMenuOpen, setIsToolsMenuOpen] = useState(false)
+      const [actionsList, setActionsList] = useState(ACTIONS_LIST)
 
       const switchToBasicModel = () => {
         const nonPremiumModel = MODELS.find(m => m.access === mojom.ModelAccess.BASIC)
         setArgs({ model: nonPremiumModel })
       }
 
+      const resetSelectedActionType = () => {
+        setSelectedActionType(undefined)
+      }
+
+      const handleActionTypeClick = (actionType: mojom.ActionType) => {
+        setSelectedActionType(actionType)
+
+        if (inputText.startsWith('/')) {
+          setInputText('')
+        }
+      }
+
+      const setInputText = (text: string) => {
+        setInputText_(text)
+
+        if (text.startsWith('/')) {
+          setIsToolsMenuOpen(true)
+
+          // effectively remove the leading slash before comparing it to the action labels.
+          const searchText = text.substring(1).toLocaleLowerCase()
+
+          const filteredList = ACTIONS_LIST.map(category => ({
+            ...category,
+            actions: category.actions.filter(action => {
+              // actionType with -1 is a subcategory so we skip
+              if (action.type === -1) return
+              return action.label.toLocaleLowerCase().includes(searchText)
+            })
+          })).filter(category => category.actions.length > 0)
+
+          setActionsList(filteredList)
+        } else {
+          setIsToolsMenuOpen(false)
+          setActionsList(ACTIONS_LIST)
+        }
+      }
+
+      const isUserSelectingActionType = () => {
```

Comment:
A good chunk of this logic is duplicated from `components/ai_chat/resources/page/state/data-context-provider.tsx` which worries me a bit - when they get out of sync it's going to be pretty confusing. Do we not have a `DataContext` in the story?

Reply:
yea, there's a lot of copy paste here. i agree it gets hard to manage this. 

no, we don't have a `DataContext` for story. do you have ideas on this?

Reply:
I'd be tempted to make a test provider, or make the existing one work in storybook.

Alternatively, you could move these methods into a custom hook exported form the `data-context-provider`.

Reply:
Ok, so I'm using the same DataContextProvider in Storybook as in the main app. However, this required mocking the page handler related APIs to avoid a Storybook runtime crash. The only caveat is that if we decide to extend the Mojom API and call it in the useEffect of DataContextProvider in the future, we will need to extend it in the mock page handler API to prevent crashes.

However, Storybook is much cleaner now. Only the state that needs to be controlled through Storybook's knobs can be passed through as an override; otherwise, other state-related functionality is self-contained and works well.